internal/tofu/eval_variable.go

```diff
@@ -477,3 +477,33 @@ You can correct this by removing references to sensitive values, or by carefully
 		FailureMessage: errorMessage,
 	}, diags
 }
+
+// evalVariableDeprecation checks a variable "deprecated" message
+func evalVariableDeprecation(addr addrs.AbsInputVariableInstance, config *configs.Variable, expr hcl.Expression, ctx EvalContext, onlyLocal bool) (diags tfdiags.Diagnostics) {
+	if !config.DeprecatedSet {
+		log.Printf("[TRACE] evalVariableDeprecation: variable %s is having no deprecation configured", addr)
+		return nil
+	}
+	if onlyLocal && strings.HasPrefix(config.DeclRange.Filename, ".terraform") {
+		log.Printf("[TRACE] evalVariableDeprecation: variable %s is part of a remote module and was requested to ignore those", addr)
+		return nil
+	}
+	// if the variable is not given in the module call, do not show a warning
+	if expr == nil {
+		log.Printf("[TRACE] evalVariableDeprecation: variable %s is marked as deprecated but is not used", addr)
+		return nil
+	}
+	val := ctx.GetVariableValue(addr)
+	if val.IsNull() {
+		log.Printf("[TRACE] evalVariableDeprecation: variable %s is marked as deprecated null value given", addr)
+		return nil
+	}
+	log.Printf("[TRACE] evalVariableDeprecation: usage of deprecated variable %q detected", addr)
+	// Any deprecated
+	return diags.Append(&hcl.Diagnostic{
+		Severity: hcl.DiagWarning,
+		Summary:  fmt.Sprintf(`The variable %q is marked as deprecated by module author`, config.Name),
+		Detail:   config.Deprecated,
```

Comment:
I missed that part in the RFC example of variable deprecation, however it is slightly different for module outputs:

```
│ This value is derived from module.mod.this_is_my_output, which is
│ deprecated with the following message:
|
| This output will be removed on 2024-12-31. Use another_output instead.
```

I believe explicitly saying this message is from the deprecation configuration (and not from OpenTofu) is a better approach. So I suggest to tweak the messaging so it is more similar to module outputs:

```
│ This variable is marked as deprecated with the following message:
|
| This variable will be removed on 2024-12-31. Use another_variable instead.
```

WDYT?

Reply:
Yep, we missed this specificity from the RFC. I totally agree that it's better to give better context to the users.
Changed in 000434c5cde89696960449ece0160595ac64c5c5
![Screenshot 2025-03-07 at 11 18 44](https://github.com/user-attachments/assets/9f54ba85-6d34-4851-88e0-e2ed20b71b27)

Reply:
Looks good, thanks!